src/test/java/io/apicurio/datamodels/LibraryTest.java

```diff
@@ -62,4 +70,34 @@ public void testReadDocumentFromJSONString() {
         Assert.assertEquals("3.0.2", ((Oas30Document) doc).openapi);
     }
 
+    @Test
+    public void testValidateWithExtensions() {
+        String jsonString = "{\r\n" +
+                "    \"openapi\": \"3.0.2\",\r\n" +
+                "    \"info\": {\r\n" +
+                "        \"title\": \"Very Simple API\",\r\n" +
+                "        \"version\": \"1.0.0\"\r\n" +
+                "    }\r\n" +
+                "}";
+        Document doc = Library.readDocumentFromJSONString(jsonString);
+
+        List<IValidationExtension> extensions = new ArrayList<>();
+        extensions.add(new TestValidationExtension());
+
+        List<ValidationProblem> problems = Library.validateWithExtensions(doc, null, extensions);
+        Assert.assertTrue(problems.stream().anyMatch(p -> p.errorCode.equals("TEST-001")));
+    }
```

Comment:
It looks like the approach using Promise won't work in Java unfortunately:

```shell
java.lang.UnsatisfiedLinkError: 'def.js.Promise def.js.Promise.resolve(java.lang.Object)'
```

https://github.com/cincheo/jsweet/issues/121#issuecomment-220381973

Reply:
I wonder - what if the method returned a `CompletableFuture`?  Perhaps JSweet is smart enough to transpile that into a `Promise`?

Reply:
Spoiler alert, JSweet does not automatically handle `CompletableFuture` as a `Promise`.  Bummer.  I'm going to see if there is a way to add a mapping for it.

Reply:
So I haven't tried it, but it looks like JSweet can be extended easily enough so that we could use `CompletableFuture` as the return type of the new validator extension interface, and have that automatically converted to `Promise`.  They support adapters like this:

https://github.com/cincheo/jsweet/blob/develop/transpiler/src/main/java/org/jsweet/transpiler/extension/BigDecimalAdapter.java

Reply:
Yeah, that is actually what I had initially tried before trying out the `Promise` candy.

Okay cool, thanks, I will give that a shot ;)

Reply:
Very cool - I think it could be a huge win in general if we can figure out how to use Adapters in JSweet.  I think there may already be some other use-cases.  One thing that leaps to mind is that I could stop using `Object` for JSON objects and instead use the Jackson class directly (`JsonNode or whatever it is).

Reply:
Yeah, I began looking at it a couple of days ago and the adapters seem straightforward enough, should get back to it on Monday :)